### PR TITLE
[#4663] Added collectibles and CK support

### LINKED ISSUE #4663: [ERC721] Add collectibles section to wallet
### User Story

As a user, I want to see my collectibles inside my Status wallet so that I can lovingly admire them and track my full portfolio of assets. 

### Description

*Type*: Feature

*Summary*: Basic list UI to display various collectible types in the wallet, beneath the user's ERC20 tokens. 

CryptoKitties and Etheremon will be the first supported collectible types.

We've seen the most demand for CryptoKitties from our community. We also know that community members own Etheremon, which is a top 20 DApp by DAU on [DApp Radar](https://dappradar.com/).

#### Expected behavior
I see a list of my CryptoKitties and Etheremon beneath my ETH, SNT and other ERC20 tokens.

Mock up:
https://www.dropbox.com/s/0224wcb81u4e6it/collectibles%201.png?dl=0

If I don't own any of these collectible types, they'll be displayed—by default—with "0" amounts. 

I should be able to manage which collectibles are shown under Manage Assets. 

#### Actual behavior
Wallet only displays ERC20 tokens.

<blockquote><div>Dropbox</div><div><strong><a href="https://www.dropbox.com/s/0224wcb81u4e6it/collectibles%201.png?dl=0">collectibles 1.png</a></strong></div><div>Shared with Dropbox</div></blockquote>

### LINKED ISSUE #4666: [ERC721] Support for CryptoKitties
### User Story

As a user, I want to see details about each CryptoKitty I own inside my Status wallet. 

### Description

*Type*: Feature

*Summary*:

- Detect how many CryptoKitties a user owns for display on the wallet main screen
- Display each CryptoKitty and its attributes on a dedicated screen

#### Expected behavior

If I own 2 CryptoKitties, I see that I have 2 CryptoKitties in my list of wallet assets.

When I click on CryptoKitties, I'm brought to a list of my collection, with details for each one:

- Kitty name
- Image
- Bio
- Last sold price? Purchase date? 
- Link to open in CryptoKitties

Mock up:
https://www.dropbox.com/s/gx7ui5yqpj1b8iq/collectibles%202.png?dl=0

#### Actual behavior
I can't see my CryptoKitties anywhere inside Status.

<blockquote><div>Dropbox</div><div><strong><a href="https://www.dropbox.com/s/gx7ui5yqpj1b8iq/collectibles%202.png?dl=0">collectibles 2.png</a></strong></div><div>Shared with Dropbox</div></blockquote>

### PR DESCRIPTION
**Pull request by jeluard**
fixes #4663
fixes #4666

### Summary:

Added wallet support for collectibles and CryptoKitties details view.

### Steps to test:
- Open Status
- Make sure you have kitties
- Browse wallet
- Enable CK asset

Note: kitty images are not supported yet due to #4709

status: ready

### CODE REVIEW

**Comment by goranjovic** [comment]
Any screenshots?

**Comment by jeluard** [comment]
@goranjovic What you don't like surprises ?

**Comment by jeluard** [comment]
nfts
ck

**Review by goranjovic — APPROVED** [review]

File: src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    :http-get-n (mapv (fn [id]
+                        {:url (str "https://api.cryptokitties.co/kitties/" id)
+                         :success-event-creator (fn [o]
+                                                  [:load-kitty-success [id (parse-payload o)]])
+                         :failure-event-creator (fn [o]
+                                                  [:load-kitty-failure [id (parse-payload o)]])})
+                      ids)}))
+
+(defn kitties-url [address]
+  (str "https://api.cryptokitties.co/kitties?offset=0&limit=12&owner_wallet_address=" address "&parents=false"))
```
Comment (flexsurfer) [inline src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs]: only 12 ?

File: src/status_im/ui/screens/wallet/collectibles/views.cljs [hunk 2]
```diff
@@ -0,0 +1,36 @@
+(ns status-im.ui.screens.wallet.collectibles.views
```
Comment (flexsurfer) [inline src/status_im/ui/screens/wallet/collectibles/views.cljs:1]: will it be extension later?

File: src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs (hunk 1 above)
Comment (jeluard) [reply src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs]: Oups good catch!

File: src/status_im/ui/screens/wallet/collectibles/views.cljs (hunk 2 above)
Comment (jeluard) [reply src/status_im/ui/screens/wallet/collectibles/views.cljs:1]: Yeahhh! :)

**Comment by jeluard** [comment]
@flexsurfer Can you accept the PR?

**Review by janherich — APPROVED** [review]

**Comment by chadyj** [comment]
Looks great! I hope in the future we can see the kitty 😺 

Any reason that "view in cryptokitties" opens in the system browser instead of the Status browser?

And how about auto showing kitties instead of the manual step? It might not be so discoverable for users otherwise.

**Comment by jeluard** [comment]
@chadyj 
> Any reason that "view in cryptokitties" opens in the system browser instead of the Status browser?

Good point! Will fix.

> And how about auto showing kitties instead of the manual step? It might not be so discoverable for users otherwise.

Do you mean skipping the `Manage assets` step? That's how we do for all assets type. Only SNT is on by default.
We could have some kind of auto-discovery feature the first time (cc @goranjovic), especially for recovery.

**Comment by chadyj** [comment]
> Do you mean skipping the Manage assets step? That's how we do for all assets type. Only SNT is on by default.

Yes exactly. 
AFAIK we are only supporting a handful of ERC721 tokens as we expect many users to have these so why not remove this hidden step for the user when it comes to ERC721? 

Eventually, we may have a `scan wallet for tokens` button or similar functionality, but for now just show kitties and SNT and other ERC721 where there is explicit support?

**Comment by jeluard** [comment]
@chadyj It's only relevant during recovery (as we create the wallet address) so we might as well transparently do the discovery.
We will have much more NFTs in the future! :sweat_smile:

**Comment by asemiankevich** [comment]
@jeluard Status shows i have 2 kitties but details show that I have 4. And i do have 4 :) 

image

image


also mocks mention about some details for particular kitty (sold price or something) 
image
are we planning to have it still? @rachelhamlin

File: src/status_im/ui/screens/db.cljs [hunk 3]
```diff
@@ -110,14 +110,17 @@
...
 (spec/def :navigation/screen-params (spec/nilable (allowed-keys :opt-un [:navigation.screen-params/network-details
                                                                          :navigation.screen-params/browser
                                                                          :navigation.screen-params/profile-qr-viewer
                                                                          :navigation.screen-params/qr-scanner
                                                                          :navigation.screen-params/group-contacts
                                                                          :navigation.screen-params/edit-contact-group
                                                                          :navigation.screen-params/dapp-description
-                                                                         :navigation.screen-params/usage-data])))
+                                                                         :navigation.screen-params/usage-data
+                                                                         :navigation.screen-params/display-collectible])))
```
Comment (rasom) [inline src/status_im/ui/screens/db.cljs:123]: 80 chars limit (it is bigger than 120 here). No jokes, it doesn't fit on my screen

File: src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  (:require [re-frame.core :as re-frame]
+            [status-im.i18n :as i18n]
+            [status-im.ui.components.action-button.action-button :as action-button]
+            [status-im.ui.components.colors :as colors]
+            [status-im.ui.components.react :as react]
+            [status-im.ui.screens.wallet.collectibles.views :as collectibles]
+            [status-im.utils.handlers :as handlers])
+  (:refer-clojure :exclude [symbol]))
+
+(def symbol :CK)
```
Comment (rasom) [inline src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs:12]: do we really want to shadow standard func? It's a bit confusing imo

File: src/status_im/ui/screens/db.cljs (hunk 3 above)
Comment (jeluard) [reply src/status_im/ui/screens/db.cljs:123]: I didn't change it much. It was already too big :sweat_smile:

**Comment by jeluard** [comment]
@asemiankevich I take a fee of two cats per tester

**Comment by asemiankevich** [comment]
how greedy you are @jeluard

**Comment by jeluard** [comment]
@asemiankevich It's only 2 cats! You still have 2!

**Comment by asemiankevich** [comment]
:okay: :(

**Comment by asemiankevich** [comment]
can you please fix that children are taken into account too? i guess this is it

File: src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs (hunk 4 above)
Comment (jeluard) [reply src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs:12]: Agree

File: src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  (js->clj (js/JSON.parse o)
+           :keywordize-keys true))
+
+(handlers/register-handler-fx
+ :load-kitties
+ (fn [{db :db} [_ ids]]
+   {:db db
+    :http-get-n (mapv (fn [id]
+                        {:url (str "https://api.cryptokitties.co/kitties/" id)
+                         :success-event-creator (fn [o]
```
Comment (rasom) [inline src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs:36]: What does `o` stay for?

**Comment by jeluard** [comment]
Children? So those are cats you own? Are they equivalent to others?

**Comment by asemiankevich** [comment]
I do own 4 cats yes.

File: src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+              :success-event-creator (fn [o]
+                                       [:load-kitties-success (map :id (:kitties (parse-payload o)))])
+              :failure-event-creator (fn [o]
+                                       [:load-collectibles-failure (parse-payload o)])
+              :timeout-ms            10000}})
+
+(defn- kitty-name [{:keys [id name]}]
+  (or name (str (i18n/label :t/cryptokitty-name) id)))
+
+(def view-style
```
Comment (rasom) [inline src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs:62]: do we want to have styles, view, fetching logic, all in a single ns?

File: src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+     [react/text {:style name-style}
+      (kitty-name m)]
+     [react/text {:number-of-lines 3
+                  :ellipsize-mode :tail}
+      bio]]]
+   [action-button/action-button {:label               (i18n/label :t/view-cryptokitties)
+                                 :icon                :icons/address
+                                 :icon-opts           {:color colors/blue}
+                                 :accessibility-label :open-collectible-button
+                                 :on-press            #(re-frame/dispatch [:open-browser {:url (str "https://www.cryptokitties.co/kitty/" id)}])}]])
```
Comment (rasom) [inline src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs:90]: 80 chars

File: src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs (hunk 6 above)
Comment (jeluard) [reply src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs:62]: It's temporary, will move to extensions once ready. So don't thing it's worth the complexity.

Comment (rasom) [reply src/status_im/ui/screens/wallet/collectibles/cryptokitties.cljs:62]: ok

**Comment by jeluard** [comment]
@asemiankevich Could you retest now? Looks like the amount is correct (10)

**Comment by asemiankevich** [comment]
@jeluard on it!

**Comment by rachelhamlin** [comment]
😅 @asemiankevich I think we'll still include last sold price if we can get it. Do we know yet @jeluard?

**Comment by jeluard** [comment]
@rachelhamlin Not with CK api only

**Comment by jeluard** [comment]
@rachelhamlin Or maybe actually but requires some investigation

**Comment by asemiankevich** [comment]
@jeluard please update the label to `(count) CryptoKitties` as discussed 🙌

**Comment by jeluard** [comment]
assets
monsters

**Comment by asemiankevich** [comment]
please merge

File: src/status_im/ui/screens/wallet/collectibles/views.cljs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+(defmulti render-collectible (fn [symbol _] symbol))
+
+(defmethod render-collectible :default [symbol {:keys [id name]}]
+  [react/view {:style styles/default-collectible}
+   [react/text (str (clojure.core/name symbol) " #" (or id name))]])
+
+(defview display-collectible []
+  (letsubs [{:keys [name symbol]} [:get-screen-params]]
+    (let [collectibles @(re-frame/subscribe [:collectibles symbol])]
```
Comment (flexsurfer) [inline src/status_im/ui/screens/wallet/collectibles/views.cljs:24]: @jeluard better to create new subscription for that

File: src/status_im/ui/screens/wallet/events.cljs [hunk 9]
```diff
@@ -99,14 +101,18 @@
  (fn [{:keys [web3 obj success-event]}]
    (ethereum/estimate-gas-web3 web3 (clj->js obj) #(re-frame/dispatch [success-event %2]))))
 
+(defn tokens-symbols [v chain]
+  (set/difference (set v) (set (map :symbol (tokens/nfts-for chain)))))
```
Comment (flexsurfer) [inline src/status_im/ui/screens/wallet/events.cljs:105]: `v`?

File: src/status_im/ui/screens/wallet/views.cljs [hunk 10]
```diff
@@ -80,21 +82,56 @@
...
-     :scroll-enabled     false
-     :key-fn             (comp str :symbol)
-     :data               assets
-     :render-fn          (render-asset currency)}]])
+(def item-icon-forward
+  [list/item-icon {:icon      :icons/forward
+                   :icon-opts {:color :gray}}])
+
+(defn- render-collectible [address-hex {:keys [symbol icon amount] :as m}]
+  (let [i        (money/to-fixed amount)
```
Comment (flexsurfer) [inline src/status_im/ui/screens/wallet/views.cljs:90]: `i` ?

File: src/status_im/ui/screens/wallet/views.cljs [hunk 11]
```diff
@@ -80,21 +82,56 @@
...
-    {:default-separator? true
-     :scroll-enabled     false
-     :key-fn             (comp str :symbol)
-     :data               assets
-     :render-fn          (render-asset currency)}]])
+(def item-icon-forward
+  [list/item-icon {:icon      :icons/forward
+                   :icon-opts {:color :gray}}])
+
+(defn- render-collectible [address-hex {:keys [symbol icon amount] :as m}]
```
Comment (flexsurfer) [inline src/status_im/ui/screens/wallet/views.cljs:89]: `m` ?

File: src/status_im/ui/screens/wallet/views.cljs [hunk 12]
```diff
@@ -80,21 +82,56 @@
...
+                     :accessibility-label (str (-> symbol name clojure.string/lower-case) "-collectible-value-text")}
+         (or i 0)]
+        [react/text {:style           styles/asset-item-currency
+                     :uppercase?      true
+                     :number-of-lines 1}
+         (name symbol)]]
+       (when details?
+         item-icon-forward)]]]))
+
+(defn group-assets [v]
```
Comment (flexsurfer) [inline src/status_im/ui/screens/wallet/views.cljs:110]: `v` ?

**Review by flexsurfer — COMMENTED** [review]
better late than never :)

**Pull request merged**